doc/source/ray-air/examples/vicuna_13b_lightning_deepspeed_finetune.ipynb

```diff
@@ -0,0 +1,1468 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!awsv2 configure set s3.max_concurrent_requests 32\n",
+    "!awsv2 configure set default.s3.preferred_transfer_client crt\n",
+    "!awsv2 configure set default.s3.target_bandwidth 100Gb/s\n",
+    "!awsv2 configure set default.s3.multipart_chunksize 8MB"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "os.system(f\"awsv2 s3 sync {result.checkpoint.uri} /mnt/local_storage/checkpoint\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The deepspeed ZeRO-3 checkpoint is a directory containing of k shards (k=16 in our case).\n",
+    "\n",
+    "- `zero_pp_rank_k_mp_rank_00_model_states.pt`: contains the model parameter skeleton of shard k.\n",
+    "- `bf16_zero_pp_rank_k_mp_rank_00_optim_states.pt`: contains the actual flattened model parameters and optimizer states of shard k.\n",
+    "\n",
+    "Next, we removed the optimizer states and consolidate the checkpoint into a single binary file using DeepSpeed utilities. Also, since we wrapped vicuna-13b within a `LightningModule`, we need to remove the prefix `_forward_module.model.model` so that we can directly load the checkpoint into a HF vicuna model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Processing zero checkpoint '/mnt/local_storage/checkpoint/model/checkpoint'\n",
+      "Detected checkpoint of type zero stage 3, world_size: 16\n",
+      "Parsing checkpoint created by deepspeed==0.9.4\n",
+      "Reconstructed Trainable fp32 state dict with 363 params 13015864320 elements\n"
+     ]
+    }
+   ],
+   "source": [
+    "import os\n",
+    "import torch\n",
+    "from deepspeed.utils.zero_to_fp32 import get_fp32_state_dict_from_zero_checkpoint\n",
+    "\n",
+    "def extract_fp32_ckpt_from_zero(zero_ckpt_dir):\n",
+    "    state_dict = get_fp32_state_dict_from_zero_checkpoint(zero_ckpt_dir)\n",
+    "    vicuna_state_dict = {\n",
+    "        k.replace(\"_forward_module.model.\", \"\"): v for k, v in state_dict.items()\n",
+    "    }\n",
+    "    torch.save(vicuna_state_dict, os.path.join(zero_ckpt_dir, \"full_model.pt\"))\n",
+    "\n",
+    "\n",
+    "full_model_ckpt_path = \"/mnt/local_storage/checkpoint/model/full_model.pt\"\n",
+    "extract_fp32_ckpt_from_zero(\"/mnt/local_storage/checkpoint/model\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Initialize Generation Pipeline\n",
+    "\n",
+    "Here, we leverage the Accelerate library to efficiently load the model onto a suitable device(GPU and CPU) and generate a HF text generation pipeline. \n",
```

Comment:
Why GPU and CPU here?

Reply:
Because the full model(26GB) cannot fit into a single A10G GPU(24GB). We have to offload some parameters to CPU.